README.md

```diff
@@ -147,10 +147,12 @@ Each rule provides a set of options for configuration.  Options with an asterisk
 | `debug` | `false` | Set to true if the rule should print debug for each operator transformation. |
 | `description` | `undefined` | Text description of the rule. |
 | `id` | `undefined` | Unique identifier for the rule. |
+| `maxRetry` | `5` | The maximum number of attempts to search for an `undefined` data layer or test the `waitUntil` predicate. |
 | `monitor` | `true` | Set to true to monitor property changes or function calls |
 | `operators` | `[]` | List of operators that transform data before a destination. |
 | `readOnLoad` | `false` | Rule-specific override for `window[‘_dlo_readOnLoad’]`. |
 | `url` | `undefined` | Specifies a regular expression that enables the rule when the page URL matches. |
+| `waitUntil` | `undefined` | Waits a desired number of milliseconds or predicate function's truthy return type before registering the rule. |
```

Comment:
I wonder if waiting a set amount of time is a necessary feature. If a customer were to use this, they'd have to anticipate timing of function executions in the browser, which can be impacted by any number of things, in which case the `waitUntil` millisecond value will be inaccurate and the rule won't register. What if you just made this a predicate function?

Reply:
One use case where the time-based delay helps is building Page Vars for A/B tests.  I'm seeing that we often have lots of these and giving it a 2-3 second delay let's the data layer populate.  The difficulty with a predicate approach is that checking if a single test exists in the data layer results in firing N-1 events.  Not knowing how many events could be in the data layer makes if difficult to write the predicate test beforehand.

The other difficulty - and this one I'm keenly interested in because it applies to a current project - is that we can't store predicate functions due to JSON unmarshalling. So we'll lose serialization if we just have it.  We could combat that problem with some pre-defined checks (e.g. `hasProperties`).  Or a safe form of [eval](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!).  If the serialization were a non-issue, I'd definitely go with just the predicate function itself.  Maybe that's the path to pursue.

Reply:
> The difficulty with a predicate approach is that checking if a single test exists in the data layer results in firing N-1 events.

Do you mean that checking to see if a single A/B test exists in the data layer results in firing N-1 events? 

Do you have an example handy? Only asking because I'm trying to build my brain on this and I'm having trouble following.

Reply:
Sent a Google Doc with a more concrete example of where we've had a need to reduce N records in a data layer to a single object payload.

Reply:
So what I gather is that all properties of the `source` will be operated on (question: does it matter if `readOnLoad` is true or false?)

I'm thinking this behavior is separate from `waitUntil` behavior. But I guess what you're suggesting is that you can apply DataLayer rules with `readOnLoad = false` and a `waitUntil` millisecond value to then operate on all of the experiment data which would have been loaded into the data layer. Do I have that right @van-fs?

Reply:
> So what I gather is that all properties of the source will be operated on ...
Correct.  And to answer your question,`readOnLoad` could be either true of false.  It influences behavior after the `waitUntil` condition has been achieved.

Conceptually, `waitUntil` should support some delayed registration/activation of a rule.  In the ideal case, we have the predicate to be an absolute indicator.  For everything else, we can fall back to a naive time-based delay.  I think having an option of last resort may be beneficial.

Reply:
> Correct. And to answer your question,readOnLoad could be either true of false. It influences behavior after the waitUntil condition has been achieved.

Ok, I think I i wasn't totally grokking `readOnLoad`. When `true` operators do their thing and destination events are called as soon as the data layer rules are applied. When `false` the operators do their thing and the destination events are called when a mutation happens for the selected property (selector defined in the rule, or inferred from the source).

Do I finally get it? :)

Reply:
Yes, exactly!  Over time the startup logic has become more sophisticated.  What we're doing now is delaying "meaningful" registration of the rule until the data layer is ready.  Originally, the notion of an `onload` event made sense:  DLO boots up, registers rules ... it should read then.  Now we see all sorts of timing issues, missing data layers, data layers that are being constructed, etc.  Maybe we could alias `readOnLoad` to be a `readOnReady` or something.  Not needed for this PR, but if the `onload` event seems opaque (and it sounds like it might be), probably need to revisit.